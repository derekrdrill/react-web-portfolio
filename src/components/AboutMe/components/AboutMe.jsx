import React from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import { Container, Row, Col } from 'react-bootstrap';
import { Typography, List, ListItem } from '@mui/material';

import { DarkLightModeContext } from '../../DarkLightMode/context/DarkLightModeContext';

import { ABOUT_ME_TEXT } from '../constants/ABOUT_ME_TEXT';
import bitmojiChillinWithBirds from '../../../assets/bitmoji_chillinWithBirds1.png';

export const AboutMe = ({ id }) => {
  // test change

  const { darkMode } = React.useContext(DarkLightModeContext);

  return (
    <>
      <PageBodyStyle darkMode={darkMode} />
      <AboutMeContainer darkMode={darkMode} id={id} fluid>
        <Row>
          <Col className='py-3'>
            <AboutMeTitle className='text-center' component='h1' darkMode={darkMode} variant='h3'>
              ABOUT ME
            </AboutMeTitle>
          </Col>
        </Row>
        <Row>
          <AboutMeTextContainer
            darkMode={darkMode}
            xs={{ span: 10, order: 'last', offset: 1 }}
            lg={{ span: 5, order: 'first', offset: 1 }}
          >
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[1].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[2].text}
            </AboutMeText>
            <List style={{ color: 'white' }}>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[3].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[4].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[5].text}</AboutMeListItem>
              <AboutMeListItem darkMode={darkMode}>{ABOUT_ME_TEXT[6].text}</AboutMeListItem>
            </List>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[7].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[8].text}
            </AboutMeText>
            <AboutMeText darkMode={darkMode} variant='subtitle1' component='p'>
              {ABOUT_ME_TEXT[9].text}
            </AboutMeText>
          </AboutMeTextContainer>
          <AboutMeBitmojiContainer xs={12} lg={6}>
            <img src={bitmojiChillinWithBirds} alt='Chillin with birds' />
          </AboutMeBitmojiContainer>
        </Row>
      </AboutMeContainer>
    </>
  );
};

const PageBodyStyle = createGlobalStyle(({ darkMode }) => ({
  body: {
    backgroundColor: darkMode ? '#030200' : 'whitesmoke',
  },
}));

export const AboutMeContainer = styled(Container)(({ darkMode }) => ({
  padding: '30px 50px 200px 50px',
  borderTop: '2px solid transparent',
  borderImage: darkMode
    ? 'linear-gradient(to right, rgba(248, 184, 255, 1), skyblue, gainsboro)'
    : 'linear-gradient(to right, rgba(176, 52, 197, 1), rgba(86, 206, 210, 1))',
  borderImageSlice: 1,
}));

export const AboutMeTitle = styled(Typography)(({ darkMode }) => ({
  fontFamily: 'Shizuru',
  fontWeight: darkMode ? 'normal' : 'bold',
  color: darkMode ? 'gainsboro' : 'darkslategray',
}));

export const AboutMeTextContainer = styled(Col)(({ darkMode }) => ({
  padding: 40,
  height: 550,
  overflowY: 'auto',
  border: '1px dashed #383838',
  borderRadius: 2,
  backgroundColor: darkMode ? '#303030' : '#eee5c1',
}));

export const AboutMeBitmojiContainer = styled(Col)({
  textAlign: 'center',
  paddingTop: 75,
  paddingBottom: 20,
});

export const AboutMeText = styled(Typography)(({ darkMode }) => ({
  fontFamily: 'Kufam',
  color: darkMode ? 'gainsboro' : '#333333',
  marginBottom: 15,
}));

export const AboutMeListItem = styled(ListItem)(({ darkMode }) => ({
  fontFamily: 'Kufam',
  color: darkMode ? 'gainsboro' : '#333333',
}));
