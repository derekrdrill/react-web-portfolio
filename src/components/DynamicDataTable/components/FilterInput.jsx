import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const FilterInput = ({ filtersDisplay, handleFilter, headerID }) =>
  filtersDisplay && <StyledInput id={headerID} className='filter' onChange={handleFilter} />;

FilterInput.propTypes = {
  filtersDisplay: PropTypes.bool,
  handleFilter: PropTypes.func,
  headerID: PropTypes.string,
};

const StyledInput = styled.input({
  height: 21,
});
