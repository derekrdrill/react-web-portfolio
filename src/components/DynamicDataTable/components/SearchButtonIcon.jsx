import React from 'react';
import PropTypes from 'prop-types';

import SearchIcon from '@mui/icons-material/Search';
import SearchOffIcon from '@mui/icons-material/SearchOff';

export const SearchButtonIcon = ({ filtersDisplay }) =>
  filtersDisplay ? <SearchOffIcon className='search-off' /> : <SearchIcon className='search-on' />;

SearchButtonIcon.propTypes = {
  filtersDisplay: PropTypes.bool,
};
