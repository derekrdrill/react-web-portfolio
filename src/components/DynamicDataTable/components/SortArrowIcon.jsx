import React from 'react';
import PropTypes from 'prop-types';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faLongArrowAltDown, faLongArrowAltUp } from '@fortawesome/fontawesome-free-solid';

export const SortArrowIcon = ({ headerID, sortColumn, sortType }) =>
  headerID === sortColumn ? (
    sortType === 'asc' ? (
      <FontAwesomeIcon className='arrow-down' icon={faLongArrowAltDown} />
    ) : sortType === 'desc' ? (
      <FontAwesomeIcon className='arrow-up' icon={faLongArrowAltUp} />
    ) : null
  ) : null;

SortArrowIcon.propTypes = {};
