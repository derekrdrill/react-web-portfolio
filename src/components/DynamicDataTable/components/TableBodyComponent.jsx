import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Checkbox, IconButton, TableBody, TableCell, TableRow } from '@mui/material';

import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import DriveFileRenameOutlineIcon from '@mui/icons-material/DriveFileRenameOutline';
import SaveIcon from '@mui/icons-material/Save';

export const TableBodyComponent = ({
  checkOneColor,
  checkedRows,
  darkMode,
  dataRows,
  editingID,
  filterColumn,
  filterValue,
  handleCheckOneRow,
  handleEditRow,
  handleRowInputChange,
  handleSaveEdits,
  handleSetDeleteRow,
  selectedRowColor,
  selectedRowFontColor,
  tableBodyColor,
}) => {
  return (
    <TableBodyStyled darkMode={darkMode}>
      {dataRows
        .filter(dataRow => dataRow[filterColumn].toUpperCase().startsWith(filterValue))
        .map(dataRow => (
          <TableBodyRow
            key={dataRow._id}
            darkMode={darkMode}
            selected={checkedRows.includes(dataRow._id)}
            selectedcolor={selectedRowColor}
            selectedfontcolor={selectedRowFontColor}
            tablebodycolor={tableBodyColor}
          >
            <TableCell>
              <Checkbox
                id={dataRow._id}
                color={darkMode ? 'primary' : checkOneColor}
                onChange={handleCheckOneRow}
                checked={checkedRows.includes(dataRow._id)}
              />
              <RemoveRecordIcon
                className='edit-delete-icon'
                id={dataRow._id}
                darkMode={darkMode}
                onClick={handleSetDeleteRow}
              >
                <DeleteForeverIcon />
              </RemoveRecordIcon>
              {editingID != dataRow._id ? (
                <EditRecordIcon
                  className='edit-delete-icon'
                  id={dataRow._id}
                  darkMode={darkMode}
                  onClick={handleEditRow}
                >
                  <DriveFileRenameOutlineIcon />
                </EditRecordIcon>
              ) : (
                <SaveRecordIcon id={dataRow._id} onClick={handleSaveEdits}>
                  <SaveIcon />
                </SaveRecordIcon>
              )}
            </TableCell>
            {Object.entries(dataRow)
              .slice(1)
              .map(dataRowInfo => (
                <TableCell key={dataRowInfo[0]}>
                  <TableBodyCellInput
                    id={dataRowInfo[0]}
                    darkMode={darkMode}
                    value={dataRowInfo[1]}
                    isediting={dataRow._id === editingID}
                    disabled={dataRow._id !== editingID}
                    onChange={handleRowInputChange}
                  />
                </TableCell>
              ))}
          </TableBodyRow>
        ))}
    </TableBodyStyled>
  );
};

TableBodyComponent.propTypes = {
  checkOneColor: PropTypes.string,
  checkedRows: PropTypes.array,
  darkMode: PropTypes.bool,
  dataRows: PropTypes.array,
  editingID: PropTypes.string,
  filterColumn: PropTypes.string,
  filterValue: PropTypes.string,
  handleCheckOneRow: PropTypes.func,
  handleEditRow: PropTypes.func,
  handleRowInputChange: PropTypes.func,
  handleSaveEdits: PropTypes.func,
  handleSetDeleteRow: PropTypes.func,
  selectedRowColor: PropTypes.string,
  selectedRowFontColor: PropTypes.string,
  tableBodyColor: PropTypes.string,
};

export const EditRecordIcon = styled(IconButton)(({ darkMode }) => ({
  color: darkMode ? '#1a1a1a' : '#2B7494',
  cursor: 'pointer',
  ':hover': {
    color: darkMode ? '#1a1a1a' : '#3C99DC',
  },
}));

export const RemoveRecordIcon = styled(IconButton)(({ darkMode }) => ({
  color: darkMode ? '#1a1a1a' : '#B22222',
  cursor: 'pointer',
  ':hover': {
    color: darkMode ? '#121212' : '#800000',
  },
}));

export const SaveRecordIcon = styled(IconButton)({
  color: 'forestgreen',
  cursor: 'pointer',
  ':hover': {
    color: 'green',
  },
});

export const TableBodyRow = styled(TableRow)(
  ({ darkMode, selectedfontcolor, selectedcolor, tablebodycolor }) => ({
    '&.Mui-selected': {
      backgroundColor: darkMode ? '#222d49' : selectedcolor,
      td: {
        input: {
          color: darkMode ? 'gainsboro' : selectedfontcolor,
        },
        '.edit-delete-icon': {
          ':hover': {
            svg: {
              color: darkMode ? 'lightgrey' : 'inherit',
            },
          },
          svg: {
            color: darkMode ? 'grey' : 'inherit',
          },
        },
      },
      ':hover': {
        backgroundColor: darkMode ? '#222d49' : selectedcolor,
      },
    },
    ':hover': {
      opacity: darkMode ? 0.9 : 0.8,
    },
    backgroundColor: darkMode ? 'grey' : tablebodycolor,
  }),
);

export const TableBodyCellInput = styled.input(({ darkMode, isediting }) => ({
  backgroundColor: isediting ? 'white' : 'transparent',
  border: isediting ? '1px grey solid' : 'none',
  borderRadius: 2,
  color: darkMode ? 'black' : 'inherit',
}));

export const TableBodyStyled = styled(TableBody)(({ darkMode }) => ({
  backgroundColor: darkMode ? 'darkgrey' : 'white',
}));
