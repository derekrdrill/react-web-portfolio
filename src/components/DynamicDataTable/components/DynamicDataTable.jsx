import React, { useState, useEffect } from 'react';
import $ from 'jquery';
import styled from 'styled-components';
import {
  TableContainer,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Checkbox,
  Button,
  Typography,
  IconButton,
  Grid,
  Alert,
  Snackbar,
} from '@mui/material';
import RotateLeftIcon from '@mui/icons-material/RotateLeft';

import { DarkLightModeContext } from '../../DarkLightMode/context/DarkLightModeContext';

import { TableBodyComponent } from './TableBodyComponent';
import { FilterInput } from './FilterInput';
import { SearchButtonIcon } from './SearchButtonIcon';
import { SortArrowIcon } from './SortArrowIcon';
import { MessageModal } from '../../Modals/MessageModal';
import { LoaderSpinner } from '../../LoaderSpinner/LoaderSpinner';

export const getSelectRowsLabel = allRowsSelected =>
  allRowsSelected ? 'Deselect all' : 'Select all';

export const getFiltersLabel = filtersDisplay => (filtersDisplay ? 'Hide filters' : 'Show filters');

export const DynamicDataTable = ({
  checkAllColor,
  checkOneColor,
  getDataRowsAPICall,
  deleteDataRowsAPICall,
  deleteDataRowNameItems,
  deleteSelectedButtonColor,
  editDataRowsAPICall,
  headers,
  selectedRowColor = 'lightcoral',
  selectedRowFontColor = 'white',
  size,
  stickyHeader = true,
  tableBodyColor = 'beige',
}) => {
  const { darkMode } = React.useContext(DarkLightModeContext);

  const [allRowsSelected, setAllRowsSelected] = useState(false);
  const [checkedRows, setCheckRows] = useState([]);
  const [dataRows, setDataRows] = useState([]);
  const [deleteModalOpen, setDeleteModalOpen] = useState(false);
  const [deleteRowID, setDeleteRowID] = useState(null);
  const [dataRowsKeys, setDataRowsKeys] = useState([]);
  const [dataRowDeleteName, setDataRowDeleteName] = useState('');
  const [editAlertOpen, setEditAlertOpen] = useState(false);
  const [filterColumn, setFilterColunmn] = useState(headers[0].headerID);
  const [filtersDisplay, setFilterDisplay] = useState(false);
  const [filterValue, setFilterValue] = useState('');
  const [editingID, setEditingID] = useState(null);
  const [editingRow, setEditingRow] = useState(null);
  const [loading, setLoading] = useState(false);
  const [sortColumn, setSortColumn] = useState(null);
  const [sortType, setSortType] = useState(null);

  const handleGetDataRowsAPICall = () => {
    fetch(getDataRowsAPICall)
      .then(response => response.json())
      .then(json => {
        setDataRows(json);
        setDataRowsKeys(Object.keys(json[0]));
      });
  };

  const handleDeleteDataRowsAPICall = deleteRowID => {
    fetch(`${deleteDataRowsAPICall}/${deleteRowID}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    });
  };

  const handleEditDataRowsAPICall = () => {
    fetch(`${editDataRowsAPICall}`, {
      method: 'POST',
      body: JSON.stringify(editingRow),
      headers: { 'Content-Type': 'application/json' },
    });
  };

  const handleDeleteModalOpen = () => setDeleteModalOpen(true);

  const handleDeleteModalClose = () => {
    setDeleteModalOpen(false);
    setDeleteRowID(null);
  };

  const handleTableReset = () => {
    setAllRowsSelected(false);
    setCheckRows([]);
    setDataRows([]);
    setEditingID(null);
    setEditingRow(null);
    setFilterColunmn(headers[0].headerID);
    setFilterDisplay(false);
    setFilterValue('');
    setSortColumn(null);
    setSortType(null);
    $('.filter').val('');

    handleGetDataRowsAPICall();
  };

  const handleTableSort = e => {
    let columnID = e.currentTarget.id;

    const sortColumn = () => {
      dataRows.sort((a, b) => {
        let sortItemA = a[columnID].toUpperCase();
        let sortItemB = b[columnID].toUpperCase();

        if (sortType === 'asc') {
          return sortItemA > sortItemB ? -1 : 1;
        } else {
          return sortItemA < sortItemB ? -1 : 1;
        }
      });
    };

    setSortColumn(columnID);
    setSortType(sortType === 'asc' ? 'desc' : 'asc');
    sortColumn();
  };

  const handleFilter = e => {
    setFilterColunmn(e.currentTarget.id);
    setFilterValue(e.target.value.toUpperCase());
  };

  const handleCheckOneRow = e => {
    let checkedRowID = e.target.id;
    let checked = e.target.checked;

    if (checked) {
      setCheckRows([...checkedRows, checkedRowID]);

      setAllRowsSelected(checkedRows.length === dataRows.length - 1);
    } else {
      setCheckRows(checkedRows.filter(arrID => arrID != checkedRowID));
      setAllRowsSelected(false);
    }
  };

  const handleCheckAllRows = e => {
    if (e.target.checked) {
      setCheckRows(dataRows.map(dataRow => dataRow._id));
      setAllRowsSelected(true);
    } else {
      setCheckRows([]);
      setAllRowsSelected(false);
    }
  };

  const handleRowInputChange = e => {
    let editColumn = e.target.id;
    let editColumnValue = e.target.value;
    let editRowIndex = dataRows.findIndex(dataRow => dataRow._id === editingID);
    let newArr = [...dataRows];

    newArr[editRowIndex][editColumn] = editColumnValue;
    setEditingRow(newArr[editRowIndex]);
    setDataRows(newArr);
  };

  const handleEditRow = e => {
    setEditingID(e.currentTarget.id);
  };

  const handleSaveEdits = () => {
    if (editingRow) {
      setEditAlertOpen(true);
    }

    handleEditDataRowsAPICall();
    setEditingID(null);
  };

  const handleEditAlertClose = () => {
    setEditAlertOpen(false);
  };

  const handleSetDeleteRow = e => {
    let pendingDeleteRow = dataRows.find(dataRow => dataRow._id === e.currentTarget.id);
    let formattedDeleteRowName = ``;

    deleteDataRowNameItems.forEach(deleteRowItem => {
      if (dataRowsKeys.includes(deleteRowItem)) {
        formattedDeleteRowName += `${pendingDeleteRow[deleteRowItem]}`;
      } else {
        formattedDeleteRowName += `${deleteRowItem}`;
      }
    });

    setDataRowDeleteName(formattedDeleteRowName);
    setDeleteRowID(e.currentTarget.id);
    handleDeleteModalOpen();
  };

  const handleDeleteSelectedRows = () => {
    handleDeleteModalClose();
    setLoading(true);

    const deleteFromDataRows = deleteRow => {
      let rowIDIndex = dataRows.findIndex(dataRow => dataRow._id === deleteRow);
      let removeDataRow = dataRows.splice(rowIDIndex, 1)[0];

      setDataRows(dataRows.filter(dataRow => dataRow != removeDataRow));

      handleDeleteDataRowsAPICall(deleteRow);
    };

    setTimeout(() => {
      setLoading(false);

      if (deleteRowID) {
        deleteFromDataRows(deleteRowID);
        return false;
      }

      checkedRows.map(checkedRow => {
        deleteFromDataRows(checkedRow);
      });
    }, 1000);
  };

  useEffect(() => handleGetDataRowsAPICall(), []);

  return (
    <TableHolder>
      <TableContainerStyled>
        <Table size={size} stickyHeader={stickyHeader}>
          <TableHead>
            <TableRow>
              <TableToolsCell darkMode={darkMode}>
                <Grid container>
                  <Grid item xs={3}>
                    <Grid container>
                      <TableToolsText variant='subtitle2' component='h1'>
                        {getSelectRowsLabel(allRowsSelected)}
                      </TableToolsText>
                    </Grid>
                    <Grid container>
                      <Checkbox
                        color={checkAllColor}
                        onChange={handleCheckAllRows}
                        checked={allRowsSelected}
                      />
                    </Grid>
                  </Grid>
                  <Grid item xs={3}>
                    <Grid container justifyContent='center'>
                      <TableToolsText variant='subtitle2' component='h1'>
                        {getFiltersLabel(filtersDisplay)}
                      </TableToolsText>
                    </Grid>
                    <Grid container justifyContent='center'>
                      <SearchButton
                        darkMode={darkMode}
                        children={<SearchButtonIcon filtersDisplay={filtersDisplay} />}
                        onClick={() => setFilterDisplay(!filtersDisplay)}
                      />
                    </Grid>
                  </Grid>
                  <Grid item xs={3}>
                    <Grid container justifyContent='center'>
                      <TableToolsText variant='subtitle2' component='h1'>
                        Reset Table
                      </TableToolsText>
                    </Grid>
                    <Grid container justifyContent='center'>
                      <ResetButton
                        darkMode={darkMode}
                        children={<RotateLeftIcon />}
                        onClick={handleTableReset}
                      />
                    </Grid>
                  </Grid>
                  <Grid item xs={3}>
                    <Grid container justifyContent='center'>
                      <TableToolsText variant='subtitle2' component='h1'>
                        Selected
                      </TableToolsText>
                    </Grid>
                    <Grid container justifyContent='center'>
                      <SelectedRowsText variant='h6' component='h1'>
                        {checkedRows.length}
                      </SelectedRowsText>
                    </Grid>
                  </Grid>
                </Grid>
              </TableToolsCell>
              {headers.map(header => (
                <TableHeadCell darkMode={darkMode} key={header.headerID}>
                  <Grid container>
                    <TableHeadCellText
                      id={header.headerID}
                      component='h1'
                      variant='body1'
                      onClick={handleTableSort}
                    >
                      {header.headerName}
                      &nbsp;
                      <SortArrowIcon
                        headerID={header.headerID}
                        sortColumn={sortColumn}
                        sortType={sortType}
                      />
                    </TableHeadCellText>
                  </Grid>
                  <Grid container>
                    <FilterInput
                      filtersDisplay={filtersDisplay}
                      handleFilter={handleFilter}
                      headerID={header.headerID}
                    />
                  </Grid>
                </TableHeadCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBodyComponent
            checkOneColor={checkOneColor}
            checkedRows={checkedRows}
            darkMode={darkMode}
            dataRows={dataRows}
            editingID={editingID}
            filterColumn={filterColumn}
            filterValue={filterValue}
            handleCheckOneRow={handleCheckOneRow}
            handleEditRow={handleEditRow}
            handleRowInputChange={handleRowInputChange}
            handleSaveEdits={handleSaveEdits}
            handleSetDeleteRow={handleSetDeleteRow}
            selectedRowColor={selectedRowColor}
            selectedRowFontColor={selectedRowFontColor}
            tableBodyColor={tableBodyColor}
          />
        </Table>
      </TableContainerStyled>
      <Button
        color={deleteSelectedButtonColor || 'warning'}
        variant='contained'
        onClick={handleDeleteModalOpen}
      >
        Delete Selected
      </Button>
      <MessageModal
        open={deleteModalOpen}
        onClose={handleDeleteModalClose}
        onSubmit={handleDeleteSelectedRows}
        title='Confirm Delete'
        body={`
                    Are you sure you want to delete 
                        ${deleteRowID ? dataRowDeleteName : 'the ' + checkedRows.length}
                        ${
                          !deleteRowID
                            ? checkedRows.length < 2
                              ? 'selected record'
                              : 'selected records'
                            : ''
                        }?
                `}
        cancelButtonText='Nevermind'
        confirmButtonText='Yup, see ya'
      />
      <Snackbar
        open={editAlertOpen}
        autoHideDuration={3000}
        anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
        onClose={handleEditAlertClose}
      >
        <Alert severity='success'>Record Updated</Alert>
      </Snackbar>
      <LoaderSpinner open={loading} />
    </TableHolder>
  );
};

export const TableHolder = styled.div({
  width: '86%',
  marginLeft: '7%',
});

export const TableContainerStyled = styled(TableContainer)({
  backgroundColor: 'white',
  borderRadius: 3,
  boxShadow: '5px 10px 25px 1px grey',
  height: 550,
  marginBottom: 15,
});

export const TableHeadCell = styled(TableCell)(({ darkMode }) => ({
  backgroundColor: darkMode ? '#292929' : 'gainsboro',
  borderLeft: 'dashed gray 1px',
}));

export const TableToolsCell = styled(TableCell)(({ darkMode }) => ({
  backgroundColor: darkMode ? '#3d3d3d' : 'lightgrey',
}));

export const TableHeadCellText = styled(Typography)({
  cursor: 'pointer',
  ':hover': {
    opacity: 0.6,
  },
});

export const TableToolsText = styled(Typography)({
  fontSize: 8,
});

export const SelectedRowsText = styled(Typography)({
  transform: 'translateY(1px)',
});

export const SearchButton = styled(IconButton)(({ darkMode }) => ({
  svg: {
    path: {
      fill: darkMode ? 'grey' : 'inherit',
    },
  },
  transform: 'translateY(1px)',
}));

export const ResetButton = styled(IconButton)(({ darkMode }) => ({
  svg: {
    path: {
      fill: darkMode ? 'grey' : 'inherit',
    },
  },
}));
