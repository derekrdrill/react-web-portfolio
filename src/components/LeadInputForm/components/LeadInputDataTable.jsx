import React from 'react';
import { DynamicDataTable } from '../../DynamicDataTable/components/DynamicDataTable';

export const headers = [
  { headerID: 'firstName', headerName: 'First Name' },
  { headerID: 'lastName', headerName: 'Last Name' },
  { headerID: 'email', headerName: 'Email' },
  { headerID: 'phone', headerName: 'Phone Number' },
];

const getLeadsAPICall = 'http://localhost:3001/leadInput';
const deleteLeadsAPICall = 'http://localhost:3001/deleteLeadInput';
const editLeadsAPICall = 'http://localhost:3001/replaceLeadInput';

export const LeadInputDataTable = () => (
  <DynamicDataTable
    checkAllColor='primary'
    checkOneColor='secondary'
    deleteDataRowsAPICall={deleteLeadsAPICall}
    deleteDataRowNameItems={['firstName', ' ', 'lastName']}
    editDataRowsAPICall={editLeadsAPICall}
    getDataRowsAPICall={getLeadsAPICall}
    headers={headers}
    size='small'
  />
);
