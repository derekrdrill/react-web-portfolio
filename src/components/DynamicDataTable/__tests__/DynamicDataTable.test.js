import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';

import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import {
  DynamicDataTable,
  ResetButton,
  SearchButton,
  TableToolsCell,
} from '../components/DynamicDataTable';

import { getFiltersLabel, getSelectRowsLabel } from '../components/DynamicDataTable';

import { headers } from '../../LeadInputForm/components/LeadInputDataTable';

configure({ adapter: new Adapter() });

const renderer = new ShallowRenderer();

describe('Dynamic data table tests', () => {
  let realUseContext;
  let useContextMock;

  beforeEach(() => {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
  });

  afterEach(() => {
    React.useContext = realUseContext;
  });

  it('renders correctly', () => {
    useContextMock.mockReturnValue({ darkMode: true });
    renderer.render(<DynamicDataTable headers={headers} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('renders TableToolsCell styled component correctly', () => {
    const tableToolsCell = shallow(<TableToolsCell />);
    const tableToolsCellDark = shallow(<TableToolsCell darkMode />);

    expect(tableToolsCell.props().className).toEqual('sc-fmKESZ kmqGbv');
    expect(tableToolsCellDark.props().className).toEqual('sc-fmKESZ eSsaJU');
  });

  it('renders SearchButton styled component correctly', () => {
    const searchButton = shallow(<SearchButton />);
    const searchButtonDark = shallow(<SearchButton darkMode />);

    expect(searchButton.props().className).toEqual('sc-dJGLCQ kdsAyp');
    expect(searchButtonDark.props().className).toEqual('sc-dJGLCQ llldCJ');
  });

  it('renders ResetButton styled component correctly', () => {
    const resetButton = shallow(<ResetButton />);
    const resetButtonDark = shallow(<ResetButton darkMode />);

    expect(resetButton.props().className).toEqual('sc-hIPCAL iSgyjz');
    expect(resetButtonDark.props().className).toEqual('sc-hIPCAL gCnygP');
  });

  it('runs getSelectRowsLabel correctly', () => {
    expect(getSelectRowsLabel(true)).toEqual('Deselect all');
    expect(getSelectRowsLabel(false)).toEqual('Select all');
  });

  it('runs getFiltersLabel correctly', () => {
    expect(getFiltersLabel(true)).toEqual('No filters');
    expect(getFiltersLabel(false)).toEqual('Filters');
  });
});
