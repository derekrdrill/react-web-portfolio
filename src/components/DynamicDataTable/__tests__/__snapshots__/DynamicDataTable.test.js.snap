// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynamic data table tests renders correctly 1`] = `
<styled.div>
  <Styled(Component)>
    <Styled(Component)
      darkMode={true}
      stickyHeader={true}
    >
      <ForwardRef(TableHead)>
        <ForwardRef(TableRow)>
          <Styled(Component)
            darkMode={true}
          >
            <ForwardRef(Grid)
              container={true}
            >
              <Styled(Component)
                component="h1"
                darkMode={true}
                id="firstName"
                onClick={[Function]}
                variant="body1"
              >
                First Name
                 
                <SortArrowIcon
                  headerID="firstName"
                  sortColumn={null}
                  sortType={null}
                />
              </Styled(Component)>
            </ForwardRef(Grid)>
            <ForwardRef(Grid)
              container={true}
            >
              <FilterInput
                filtersDisplay={false}
                handleFilter={[Function]}
                headerID="firstName"
              />
            </ForwardRef(Grid)>
          </Styled(Component)>
          <Styled(Component)
            darkMode={true}
          >
            <ForwardRef(Grid)
              container={true}
            >
              <Styled(Component)
                component="h1"
                darkMode={true}
                id="lastName"
                onClick={[Function]}
                variant="body1"
              >
                Last Name
                 
                <SortArrowIcon
                  headerID="lastName"
                  sortColumn={null}
                  sortType={null}
                />
              </Styled(Component)>
            </ForwardRef(Grid)>
            <ForwardRef(Grid)
              container={true}
            >
              <FilterInput
                filtersDisplay={false}
                handleFilter={[Function]}
                headerID="lastName"
              />
            </ForwardRef(Grid)>
          </Styled(Component)>
          <Styled(Component)
            darkMode={true}
          >
            <ForwardRef(Grid)
              container={true}
            >
              <Styled(Component)
                component="h1"
                darkMode={true}
                id="email"
                onClick={[Function]}
                variant="body1"
              >
                Email
                 
                <SortArrowIcon
                  headerID="email"
                  sortColumn={null}
                  sortType={null}
                />
              </Styled(Component)>
            </ForwardRef(Grid)>
            <ForwardRef(Grid)
              container={true}
            >
              <FilterInput
                filtersDisplay={false}
                handleFilter={[Function]}
                headerID="email"
              />
            </ForwardRef(Grid)>
          </Styled(Component)>
          <Styled(Component)
            darkMode={true}
          >
            <ForwardRef(Grid)
              container={true}
            >
              <Styled(Component)
                component="h1"
                darkMode={true}
                id="phone"
                onClick={[Function]}
                variant="body1"
              >
                Phone Number
                 
                <SortArrowIcon
                  headerID="phone"
                  sortColumn={null}
                  sortType={null}
                />
              </Styled(Component)>
            </ForwardRef(Grid)>
            <ForwardRef(Grid)
              container={true}
            >
              <FilterInput
                filtersDisplay={false}
                handleFilter={[Function]}
                headerID="phone"
              />
            </ForwardRef(Grid)>
          </Styled(Component)>
        </ForwardRef(TableRow)>
      </ForwardRef(TableHead)>
      <TableBodyComponent
        checkedRows={Array []}
        darkMode={true}
        dataRows={Array []}
        editingID={null}
        filterColumn="firstName"
        filterValue=""
        handleCheckOneRow={[Function]}
        handleEditRow={[Function]}
        handleRowInputChange={[Function]}
        handleSaveEdits={[Function]}
        handleSetDeleteRow={[Function]}
        selectedRowColor="lightcoral"
        selectedRowFontColor="black"
        tableBodyColor="beige"
      />
    </Styled(Component)>
  </Styled(Component)>
  <MessageModal
    body="
                    Are you sure you want to delete 
                        the 0
                        selected record?
                "
    cancelButtonText="Nevermind"
    confirmButtonText="Yup, see ya"
    onClose={[Function]}
    onSubmit={[Function]}
    open={false}
    title="Confirm Delete"
  />
  <ForwardRef(Snackbar)
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={3000}
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(Alert)
      severity="success"
    >
      Record Updated
    </ForwardRef(Alert)>
  </ForwardRef(Snackbar)>
  <LoaderSpinner
    open={false}
  />
</styled.div>
`;
